### DISCUSSION TITLE
Maplibre ignores CapacitorHttp implementation

### DISCUSSION THREAD

**Discussion by Duoquote** [maplibre/maplibre-gl-js#5250] [Q&A] [answered]
**maplibre-gl-js version**:  4.7.1

**browser**: Mozilla/5.0 (Linux; Android 13; SM-G780G Build/TP1A.220624.014; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/131.0.6778.135 Mobile Safari/537.36

**Context**: When using [CapacitorJS](https://capacitorjs.com/), you can enabled so called `CapacitorHttp` plugin, which is something like a wrapper/override over XMLHttpRequest and Fetch apis. It patches and forwards these requests to internal handler on android/ios and executes requests internally, avoiding using WebView's http implementation. When requesting to an external source, the WebView as you can see blocks the requests because of CORS, there is really no easy way to get around this so by using CapacitorHttp you can avoid CORS.

![Image](https://github.com/user-attachments/assets/32d9993b-4b65-4e36-b684-91e3636f36db)

Example request that goes through CapacitorHttp;
![Image](https://github.com/user-attachments/assets/7f090318-2205-4de2-8a85-a4018e435903)

So it's something like a proxy within the app.

Example request using CapacitorHttp ([source](https://capacitorjs.com/docs/apis/http#example));

```js
import { CapacitorHttp } from '@capacitor/core';

// Example of a GET request
const doGet = () => {
  const options = {
    url: 'https://example.com/my/api',
    headers: { 'X-Fake-Header': 'Fake-Value' },
    params: { size: 'XL' },
  };

  const response: HttpResponse = await CapacitorHttp.get(options);

  // or...
  // const response = await CapacitorHttp.request({ ...options, method: 'GET' })
};

// Example of a POST request. Note: data
// can be passed as a raw JS Object (must be JSON serializable)
const doPost = () => {
  const options = {
    url: 'https://example.com/my/api',
    headers: { 'X-Fake-Header': 'Fake-Value' },
    data: { foo: 'bar' },
  };

  const response: HttpResponse = await CapacitorHttp.post(options);

  // or...
  // const response = await CapacitorHttp.request({ ...options, method: 'POST' })
};
```

My main question is, can I in some way send requests through `CapacitorHttp` interface? I have seen there is [transformRequest](https://maplibre.org/maplibre-gl-js/docs/API/type-aliases/MapOptions/#transformrequest) but that just really lets you modify only the headers, I am actually already using it to pass authorization header;

```js
      gl_ = Leaflet.maplibreGL({
        style: `${host}/v4/maps/`,
        pane: "maplibre-gl",
        transformRequest: (url, resourceType) => {
          if (!url.startsWith(host)) return { url };

          return ["Tile", "Style"].includes(resourceType) ? {
            url,
            credentials: "include",
            headers: {
              "Authorization": `JWT ${token}`,
            },
          } : { url };
        },
      });
```

Also, why is CapacitorHttp's patch is not working on maplibre-gl?

**Comment by HarelM** [comment]
If capacitor doesn't psych correctly the fetch and xhr of the browser, then it's a question for capacitor, it might be related to cuddle running in worker thread if I needed to guess. 
You can use addProtocol to overcome these issues I belive.

**Reply by Duoquote** [reply]
Yeah I also found about 30 minutes later that the issue is indeed in workers as Capacitor is not loading it's code into workers, I am trying some things and will let you know about the progress.

**Comment by Duoquote** [comment, marked as answer]
### Okay I just tried some hack while writing this post, solution is at the bottom.

I have been tinkering with this for a while now, looks like I am stuck in a order of compatibility problems;

Packages;
- Vite.js
- Capacitor.js
- Maplibre-gl

Given this worker file (not working, just demonstration);
```js

// import { CapacitorHttp } from "@capacitor/core";
// importScripts("/@capacitor/core");

var CapacitorHttp;
import("@capacitor/core").then(module => {
  CapacitorHttp = module.CapacitorHttp;
});

async function loadFn(params, abortController) {
  const urlNext = params.url.split("://")[1];
  const response = await CapacitorHttp.get({
    url: `${urlNext.startsWith("localhost") ? "http" : "https"}://${urlNext}`,
    headers: params.headers,
    params: params.params,
  })
  console.log("Worker", params, response);
  if (response.headers["content-type"] === "application/x-protobuf" && typeof response.data === "string") {
    // base64ToBlob(response.data, "application/x-protobuf");
    const bytes = new Uint8Array([...response.data].map(char => char.charCodeAt(0)));

    // Now create a Blob from these bytes:
    const blob = new Blob([bytes], { type: "application/x-protobuf" });
    return {
      data: await blob.arrayBuffer(),
    }
  }
  return response;
}

self.addProtocol("capacitor", loadFn);
```

# Situation 1:
Using `import workerUrl from "../workers/map-worker?worker&url"` and `importScriptInWorkers(new URL(workerUrl, import.meta.url).href)` results in;

![image](https://github.com/user-attachments/assets/9cc8a014-f174-4191-a47b-84d1ca85b570)

Which I need to run worker with `type: module` (which I can't) results in;

![image](https://github.com/user-attachments/assets/54d445e0-990a-4387-a831-4d6e38781b46)

Imports look fine but it uses `import` statements obviously;

![image](https://github.com/user-attachments/assets/9906b6cb-d979-4df8-a4e7-6cc66b730ced)

# Situation 2:
With some hacky usage I tried (according to previous results);
```js
importScriptInWorkers(new URL(workerUrl.replace("module", "classic"), import.meta.url).href)
```

But it's not really helping;

```js
Uncaught (in promise) SyntaxError: Failed to execute 'importScripts' on 'WorkerGlobalScope': The URL '/@vite/env' is invalid.
```

# Situation 3:
Vite [docs mention](https://vite.dev/guide/features.html#web-workers) that this specific syntax should import, but I can't use `Worker` object with `importScriptInWorkers`, maybe if I just try it works?

```js
importScriptInWorkers(new URL("../workers/map-worker.js", import.meta.url));
```
![image](https://github.com/user-attachments/assets/9f01a7d8-ec14-484c-88bc-f84664885680)

This does not work, let's try using `.href` method;
```js
importScriptInWorkers(new URL("../workers/map-worker.js", import.meta.url).href)
```

Okay it loads, but now, how do I import `CapacitorHttp` from `@capacitor/core`? Maybe if I just import it converts to what it's supposed to be? It doesn't.

![image](https://github.com/user-attachments/assets/c9a09c73-f3f0-4ece-b1ca-fe369fdd3f27)
![image](https://github.com/user-attachments/assets/79edddb0-d889-4ab7-afcc-f04b5c0281c5)

# Looks like solved?

Using `import` function by itself I was able to import the package. There is an error waiting for me but that's on capacitor side, I will just leave the post as is as if someone encounters same things at least there is a thought process they can follow.

Here is the final state of worker script;
```js
var CapacitorHttp;
import("@capacitor/core").then(module => {
  CapacitorHttp = module.CapacitorHttp;
});

async function loadFn(params, abortController) {
  const urlNext = params.url.split("://")[1];
  const response = await CapacitorHttp.get({
    url: `${urlNext.startsWith("localhost") ? "http" : "https"}://${urlNext}`,
    headers: params.headers,
    params: params.params,
  })
  console.log("Worker", params, response);
  if (response.headers["content-type"] === "application/x-protobuf" && typeof response.data === "string") {
    // base64ToBlob(response.data, "application/x-protobuf");
    const bytes = new Uint8Array([...response.data].map(char => char.charCodeAt(0)));

    // Now create a Blob from these bytes:
    const blob = new Blob([bytes], { type: "application/x-protobuf" });
    return {
      data: await blob.arrayBuffer(),
    }
  }
  return response;
}

self.addProtocol("capacitor", loadFn);
```

And in my project;
```js
import { importScriptInWorkers } from "maplibre-gl";

...
      gl_ = Leaflet.maplibreGL({
        style: `${host}/v4/maps/`,
        pane: "maplibre-gl",
        transformRequest: (url, resourceType) => {
          return ["Tile", "Style"].includes(resourceType) ? {
            // This is important;
            url: url.replace(/^https?/, "capacitor"),
            credentials: "include",
            headers: {
              "Authorization": `JWT ${token}`,
            },
          } : { url };
        },
      });
      
      importScriptInWorkers(new URL("../workers/map-worker.js", import.meta.url).href);
```

![image](https://github.com/user-attachments/assets/613d9c99-a625-4216-8e80-697a08d0bffe)

**Comment by HarelM** [comment]
I think this is the first time I saw someone achieving to use addProtocol in the worker context.
To be honest, when I wrote it I was not sure it will ever work :-) Kudos!

**Reply by Duoquote** [reply]
Is it possible to disable worker logic altogether? Or can I create my own tile loader and load vector tiles on map manually? I am experiencing lots of compatibility problem across sentry/maplibre-gl/vite/capacitor modules on mobile devices.

**Reply by Duoquote** [reply]
Maybe I can create a custom source by overriding `loadTile` function in;

https://github.com/maplibre/maplibre-gl-js/blob/47ed30d6ec105cbfb8768fd27da1b317af31b903/src/source/vector_tile_source.ts#L191

And using functions;
https://github.com/maplibre/maplibre-gl-js/blob/8f1f5f71644f74d05abfe52a3f73d3bfc8d62f59/src/source/vector_tile_worker_source.ts#L95
https://github.com/maplibre/maplibre-gl-js/blob/8f1f5f71644f74d05abfe52a3f73d3bfc8d62f59/src/source/vector_tile_worker_source.ts#L67

**Reply by HarelM** [reply]
Just use `addProtocol` on the main thread without using it in the worker thread.
Requests in the worker that use `addPrtotocol` are routed to the main thread and you can handle them there with capacitor "regularly".

**Reply by Duoquote** [reply]
Oh okay, I thought `addProtocol` was made for workers only

**Reply by Duoquote** [reply]
Finally got it working with this configuration;

```js
      addProtocol("capacitor", async (params, abortController) => {
        const url = params.url.replace("capacitor://", params.ssl ? "https://" : "http://");
        if (deviceInfo.platform === "web") {
          return await axios({
            method: "GET",
            url,
            responseType: "arraybuffer",
          }).then(res => ({ data: res.data }));
        } else {
          return await CapacitorHttp.request({
            url,
            method: "GET",
            headers: params.headers,
          }).then(res => ({ data: Buffer.from(res.data, "base64").buffer }));
        }
      })
```

Thanks!
